### PR TITLE
Allow connectable blocks to choose what they can connect to

### PR DESCRIPTION
**Pull request by LogicTechCorp**
1.11 version of #3354

### CODE REVIEW

**Comment by williewillus** [comment]
Why the separate bouncer methods for each block?
Also if we're gonna patch the lines with 6 ternaries might as well separate it and make it readable

**Comment by LogicTechCorp** [comment]
@williewillus 
What do you mean by bouncer methods?
Which lines are you referring to?

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 1]
```diff
@@ -1380,12 +1380,25 @@
...
 +
++    /**
++     * Determines if another block can connect to this block
++     *
++     * @param world The current world
++     * @param pos The position of this block
++     * @param facing The side the connecting block is on
++     * @return True to allow another block to connect to this block
++     */
++    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (liach) [inline patches/minecraft/net/minecraft/block/Block.java.patch:1357]: `canConnectTo`

**Comment by HenryLoenwind** [comment]
@williewillus: because @mezz had LogicTechCorp change it from one big check in Block to the logic spread out to each block.

File: patches/minecraft/net/minecraft/block/Block.java.patch (hunk 1 above)
Comment (LogicTechCorp) [reply patches/minecraft/net/minecraft/block/Block.java.patch:1357]: I would call it `canConnectTo` but it is not _**what can this connect to**_ but _**what can connect to this**_.

**Comment by LogicTechCorp** [comment]
Conflicts have been resolved. 
@mezz do you have any more suggestions?

**Comment by mezz** [comment]
I think it looks good but @williewillus should clarify what they mean since nobody seems to understand and it could be something to improve the PR.

**Comment by LogicTechCorp** [comment]
@williewillus Can you clarify what you mean?

File: patches/minecraft/net/minecraft/block/Block.java.patch (hunk 1 above)
Comment (liach) [reply patches/minecraft/net/minecraft/block/Block.java.patch:1357]: `acceptsConnection`

**Comment by mezz** [comment]
I think what he's saying is that instead of making lines like this:
`return p_176221_1_.func_177226_a(field_176526_a, Boolean.valueOf(canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.NORTH))).func_177226_a(field_176525_b, Boolean.valueOf(canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.EAST))).func_177226_a(field_176527_M, Boolean.valueOf(canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.SOUTH))).func_177226_a(field_176528_N, Boolean.valueOf(canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.WEST)));`

You should make a new private method and do all the logic there where it can be formatted more nicely.

**Comment by williewillus** [comment]
Oh yeah, that wasn't 6 ternaries, that was 6 `withProperty`'s, my bad.
It doesn't need to be a separate method but just separating it onto different lines would make it more readable, since we're patching it anyways might as well.

**Comment by williewillus** [comment]
And you can also remove the explicit boxing (Boolean.valueOf)

**Comment by LogicTechCorp** [comment]
@williewillus Working on that now.

**Comment by LogicTechCorp** [comment]
Is there anything else you would like me to change? @williewillus @mezz

**Comment by mezz** [comment]
You mostly have the right idea, but you should move things around so that the patches to existing methods are 1 line only.

So you need a new private method at the bottom that does all this:
```
++        return p_176221_1_.func_177226_a(field_176526_a, canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.NORTH))
 ++                .func_177226_a(field_176525_b, canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.EAST))
 ++                .func_177226_a(field_176527_M, canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.SOUTH))
 ++                .func_177226_a(field_176528_N, canFenceConnectTo(p_176221_2_, p_176221_3_, EnumFacing.WEST));
```

and your patch to `func_176221_a` would just replace the single line with a call to your new method.

**Comment by LogicTechCorp** [comment]
Does it look good now? @mezz

File: patches/minecraft/net/minecraft/block/BlockFenceGate.java.patch [hunk 2]
```diff
@@ -0,0 +1,39 @@
+--- ../src-base/minecraft/net/minecraft/block/BlockFenceGate.java
++++ ../src-work/minecraft/net/minecraft/block/BlockFenceGate.java
+@@ -48,11 +48,11 @@
+ 
+     public IBlockState func_176221_a(IBlockState p_176221_1_, IBlockAccess p_176221_2_, BlockPos p_176221_3_)
+     {
+-        EnumFacing.Axis enumfacing$axis = ((EnumFacing)p_176221_1_.func_177229_b(field_185512_D)).func_176740_k();
++        EnumFacing.Axis axis = p_176221_1_.func_177229_b(field_185512_D).func_176740_k();
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockFenceGate.java.patch]: this line should not be changed

File: patches/minecraft/net/minecraft/block/BlockFenceGate.java.patch [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+ 
+-        if (enumfacing$axis == EnumFacing.Axis.Z && (p_176221_2_.func_180495_p(p_176221_3_.func_177976_e()).func_177230_c() == Blocks.field_150463_bK || p_176221_2_.func_180495_p(p_176221_3_.func_177974_f()).func_177230_c() == Blocks.field_150463_bK) || enumfacing$axis == EnumFacing.Axis.X && (p_176221_2_.func_180495_p(p_176221_3_.func_177978_c()).func_177230_c() == Blocks.field_150463_bK || p_176221_2_.func_180495_p(p_176221_3_.func_177968_d()).func_177230_c() == Blocks.field_150463_bK))
++        if (axis == EnumFacing.Axis.Z && (canFenceGateConnectTo(p_176221_2_, p_176221_3_, EnumFacing.WEST) || canFenceGateConnectTo(p_176221_2_, p_176221_3_, EnumFacing.EAST)) || axis == EnumFacing.Axis.X && (canFenceGateConnectTo(p_176221_2_, p_176221_3_, EnumFacing.NORTH) || canFenceGateConnectTo(p_176221_2_, p_176221_3_, EnumFacing.SOUTH)))
+         {
+-            p_176221_1_ = p_176221_1_.func_177226_a(field_176467_M, Boolean.valueOf(true));
++            p_176221_1_ = p_176221_1_.func_177226_a(field_176467_M, true);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockFenceGate.java.patch]: this line should not be changed

File: patches/minecraft/net/minecraft/block/BlockFence.java.patch [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
++        return false;
++    }
++
++    public boolean canFenceConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
++    {
++        Block block = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return block.canBeConnectedTo(world, pos.func_177972_a(facing), facing.func_176734_d()) || func_176524_e(world, pos.func_177972_a(facing));
++    }
++
++    public IBlockState getFenceState(IBlockState state, IBlockAccess world, BlockPos pos)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockFence.java.patch]: make this private or protected

File: patches/minecraft/net/minecraft/block/BlockFence.java.patch [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
++            else if(this == Blocks.field_150386_bk && connector != Blocks.field_150386_bk)
++            {
++                return false;
++            }
++            return true;
++        }
++        return false;
++    }
++
++    public boolean canFenceConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockFence.java.patch]: make this private or protected

File: patches/minecraft/net/minecraft/block/BlockFenceGate.java.patch [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
++    /* ======================================== FORGE START ======================================== */
++
++    @Override
++    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
++    {
++        Block connector = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return connector instanceof BlockFence || connector instanceof BlockWall;
++    }
++
++    public boolean canFenceGateConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockFenceGate.java.patch]: make this private or protected

File: patches/minecraft/net/minecraft/block/BlockPane.java.patch [hunk 7]
```diff
@@ -1,26 +1,41 @@
...
++    /* ======================================== FORGE START ======================================== */
++
++    @Override
++    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
++    {
++        Block connector = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return connector instanceof BlockPane;
++    }
++
++    public boolean canPaneConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockPane.java.patch]: make this private or protected

File: patches/minecraft/net/minecraft/block/BlockPane.java.patch [hunk 8]
```diff
@@ -1,26 +1,41 @@
...
 +    {
-+        BlockPos off = pos.func_177972_a(dir);
-+        IBlockState state = world.func_180495_p(off);
-+        return func_150098_a(state.func_177230_c()) || state.isSideSolid(world, off, dir.func_176734_d());
++        IBlockState state = world.func_180495_p(pos.func_177972_a(facing));
++        return state.func_177230_c().canBeConnectedTo(world, pos.func_177972_a(facing), facing.func_176734_d()) || func_150098_a(state.func_177230_c()) || state.isSideSolid(world, pos.func_177972_a(facing), facing.func_176734_d());
 +    }
++
++    public IBlockState getPaneState(IBlockState state, IBlockAccess world, BlockPos pos)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockPane.java.patch]: make this private or protected

File: patches/minecraft/net/minecraft/block/BlockWall.java.patch [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
++    /* ======================================== FORGE START ======================================== */
++
++    @Override
++    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
++    {
++        Block connector = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return connector instanceof BlockWall || connector instanceof BlockFenceGate;
++    }
++
++    public boolean canWallConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockWall.java.patch]: make this private or protected

File: patches/minecraft/net/minecraft/block/BlockWall.java.patch [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
++    {
++        Block block = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return block.canBeConnectedTo(world, pos.func_177972_a(facing), facing.func_176734_d()) || func_176253_e(world, pos.func_177972_a(facing));
++    }
++
++    public IBlockState getWallState(IBlockState state, IBlockAccess world, BlockPos pos, boolean north, boolean east, boolean south, boolean west, boolean up)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockWall.java.patch]: make this private or protected

**Review by mezz — CHANGES_REQUESTED** [review]
Ok looking good, some final requests. The helper methods I marked should be private if nothing should override them, or protected otherwise. Outside code does not need to see them or call them so they should not be public.

**Comment by LogicTechCorp** [comment]
Okay, all changes have been implemented. @mezz

File: patches/minecraft/net/minecraft/block/BlockPane.java.patch [hunk 11]
```diff
@@ -1,26 +1,41 @@
...
++    /* ======================================== FORGE START ======================================== */
++
++    @Override
++    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
++    {
++        Block connector = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return connector instanceof BlockPane;
++    }
++
++    private boolean canPaneConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockPane.java.patch]: Sorry I made a mistake, this existing method must be kept `public` to ensure compatibility with old code. I didn't realize it was here already.

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by LogicTechCorp** [comment]
`canPaneConnectTo` has been made public. @mezz

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Looks good, thanks for your patience. I'll test it out now.

**Comment by mezz** [comment]
Tested vanilla vs this patch and all connections appear to be working correctly still.

File: patches/minecraft/net/minecraft/block/BlockFence.java.patch [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
++        return false;
++    }
++
++    private boolean canFenceConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
++    {
++        Block block = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return block.canBeConnectedTo(world, pos.func_177972_a(facing), facing.func_176734_d()) || func_176524_e(world, pos.func_177972_a(facing));
++    }
++
++    private IBlockState getFenceState(IBlockState state, IBlockAccess world, BlockPos pos)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockFence.java.patch]: Lex says you can move this code into `func_176221_a` even though it's a 4 line long patch. He does not want this helper function.

File: patches/minecraft/net/minecraft/block/BlockPane.java.patch [hunk 13]
```diff
@@ -1,26 +1,41 @@
...
      {
 -        return p_176221_1_.func_177226_a(field_176241_b, Boolean.valueOf(this.func_150098_a(p_176221_2_.func_180495_p(p_176221_3_.func_177978_c()).func_177230_c()))).func_177226_a(field_176243_N, Boolean.valueOf(this.func_150098_a(p_176221_2_.func_180495_p(p_176221_3_.func_177968_d()).func_177230_c()))).func_177226_a(field_176244_O, Boolean.valueOf(this.func_150098_a(p_176221_2_.func_180495_p(p_176221_3_.func_177976_e()).func_177230_c()))).func_177226_a(field_176242_M, Boolean.valueOf(this.func_150098_a(p_176221_2_.func_180495_p(p_176221_3_.func_177974_f()).func_177230_c())));
-+        return p_176221_1_.func_177226_a(field_176241_b, canPaneConnectTo(p_176221_2_, p_176221_3_, EnumFacing.NORTH))
-+                .func_177226_a(field_176243_N, canPaneConnectTo(p_176221_2_, p_176221_3_, EnumFacing.SOUTH))
-+                .func_177226_a(field_176244_O, canPaneConnectTo(p_176221_2_, p_176221_3_, EnumFacing.WEST))
-+                .func_177226_a(field_176242_M, canPaneConnectTo(p_176221_2_, p_176221_3_, EnumFacing.EAST));
++        return getPaneState(p_176221_1_, p_176221_2_, p_176221_3_);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockPane.java.patch]: Lex says you should put this line back the way it was before. He does not want the `getPaneState` helper function.

File: patches/minecraft/net/minecraft/block/BlockWall.java.patch [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+-        boolean flag4 = flag && !flag1 && flag2 && !flag3 || !flag && flag1 && !flag2 && flag3;
+-        return p_176221_1_.func_177226_a(field_176256_a, Boolean.valueOf(!flag4 || !p_176221_2_.func_175623_d(p_176221_3_.func_177984_a()))).func_177226_a(field_176254_b, Boolean.valueOf(flag)).func_177226_a(field_176257_M, Boolean.valueOf(flag1)).func_177226_a(field_176258_N, Boolean.valueOf(flag2)).func_177226_a(field_176259_O, Boolean.valueOf(flag3));
++        boolean canConnectN = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.NORTH);
++        boolean canConnectE = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.EAST);
++        boolean canConnectS = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.SOUTH);
++        boolean canConnectW = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.WEST);
++        boolean canConnectU = canConnectN && !canConnectE && canConnectS && !canConnectW || !canConnectN&& canConnectE && !canConnectS && canConnectW;
++        return getWallState(p_176221_1_, p_176221_2_, p_176221_3_, canConnectN, canConnectE, canConnectS, canConnectW, canConnectU);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockWall.java.patch]: Lex says you should put the `getWallState` helper function back in here like it was before, and do not rename the local variables (flags).

**Review by mezz — CHANGES_REQUESTED** [review]
Lex requests some changes before this can be accepted

**Comment by LogicTechCorp** [comment]
Are these correct? @mezz

File: patches/minecraft/net/minecraft/block/BlockPane.java.patch [hunk 15]
```diff
@@ -12,15 +12,25 @@
...
++    public boolean canBeConnectedTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
 +    {
-+        BlockPos off = pos.func_177972_a(dir);
-+        IBlockState state = world.func_180495_p(off);
-+        return func_150098_a(state.func_177230_c()) || state.isSideSolid(world, off, dir.func_176734_d());
++        Block connector = world.func_180495_p(pos.func_177972_a(facing)).func_177230_c();
++        return connector instanceof BlockPane;
 +    }
++
++    public boolean canPaneConnectTo(IBlockAccess world, BlockPos pos, EnumFacing facing)
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockPane.java.patch]: do not rename the parameters here, it should match the old ones:
`public boolean canPaneConnectTo(IBlockAccess world, BlockPos pos, EnumFacing dir)`

File: patches/minecraft/net/minecraft/block/BlockWall.java.patch [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+-        boolean flag1 = this.func_176253_e(p_176221_2_, p_176221_3_.func_177974_f());
+-        boolean flag2 = this.func_176253_e(p_176221_2_, p_176221_3_.func_177968_d());
+-        boolean flag3 = this.func_176253_e(p_176221_2_, p_176221_3_.func_177976_e());
++        boolean flag = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.NORTH);
++        boolean flag1 = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.EAST);
++        boolean flag2 = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.SOUTH);
++        boolean flag3 = canWallConnectTo(p_176221_2_, p_176221_3_, EnumFacing.WEST);
+         boolean flag4 = flag && !flag1 && flag2 && !flag3 || !flag && flag1 && !flag2 && flag3;
+-        return p_176221_1_.func_177226_a(field_176256_a, Boolean.valueOf(!flag4 || !p_176221_2_.func_175623_d(p_176221_3_.func_177984_a()))).func_177226_a(field_176254_b, Boolean.valueOf(flag)).func_177226_a(field_176257_M, Boolean.valueOf(flag1)).func_177226_a(field_176258_N, Boolean.valueOf(flag2)).func_177226_a(field_176259_O, Boolean.valueOf(flag3));
++        return getWallState(p_176221_1_, p_176221_2_, p_176221_3_, flag, flag1, flag2, flag3, flag4);
```
Comment (mezz) [inline patches/minecraft/net/minecraft/block/BlockWall.java.patch]: can you replace this line with the vanilla line that was there before and remove `getWallState`, or are they different?

**Review by mezz — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/block/BlockWall.java.patch (hunk 16 above)
Comment (LogicTechCorp) [reply patches/minecraft/net/minecraft/block/BlockWall.java.patch]: They are the same, will change.

**Comment by LogicTechCorp** [comment]
Implemented changes. @mezz

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Looks good.

**Pull request merged**